R/propose_snapshot.R

```diff
@@ -0,0 +1,55 @@
+#' @title Propose snapshot
+#' @export
+#' @family staging
+#' @description Propose a Production snapshot of Staging.
+#' @details [recommend_snapshot()] proposes a snapshot of Staging
+#'   to migrate to Production. The recommended snapshot is the list of
+#'   packages for which (1) the build and check results of the current
+#'   release are in Staging, and (2) there are no issues.
+#' @return `NULL` (invisibly)
+#' @inheritParams update_staging
+#' @param repo_staging Character string, URL of the staging universe.
+#' @examples
+#' \dontrun{
+#' url_staging = "https://github.com/r-multiverse/staging"
+#' path_staging <- tempfile()
+#' gert::git_clone(url = url_staging, path = path_staging)
+#' propose_snapshot(
+#'   path_staging = path_staging,
+#'   repo_staging = "https://staging.r-multiverse.org"
+#' )
+#' }
+propose_snapshot <- function(
+  path_staging,
+  repo_staging = "https://staging.r-multiverse.org",
+  mock = NULL
+) {
+  issues <- list.files(
+    file.path(path_staging, "issues"),
+    all.files = TRUE,
+    no.. = TRUE
+  )
+  file_staging <- file.path(path_staging, "packages.json")
+  json_staging <- jsonlite::read_json(file_staging, simplifyVector = TRUE)
+  json_staging <- json_staging[, c("package", "url", "branch")]
+  meta_staging <- mock$staging %||% meta_packages(repo_staging)
+  meta_staging <- meta_staging[, c("package", "remotesha")]
+  staging <- merge(
+    x = json_staging,
+    y = meta_staging,
+    all.x = FALSE,
+    all.y = FALSE
+  )
+  staging <- staging[staging$branch == staging$remotesha,, drop = FALSE] # nolint
+  staging <- staging[!(staging$package %in% issues),, drop = FALSE] # nolint
+  staging$remotesha <- NULL
+  file_snapshot <- file.path(path_staging, "snapshot.json")
+  jsonlite::write_json(staging, file_snapshot, pretty = TRUE)
+  url <- paste0(
+    "https://staging.r-multiverse.org/api/snapshot/zip",
+    "?types=src,win,mac,linux,wasm,docs&packages=",
```

Comment:
From https://github.com/r-multiverse/help/issues/78#issuecomment-2275972225 we should probably limit binaries to win and mac for r-release.
The API provides for a specific R version so we'd need to add an argument to pass through to `&binaries=` in the URL.
We could add 'types' and 'binaries' as function arguments to afford us maximum flexibility, but with the correct defaults set.

Reply:
`binaries` is to select the versions of R, and it already automatically includes current and previous. Just added a `types` argument with default `c("win", "mac")`.

Reply:
According to Jeroen's comment, we might only want to include binaries for one version of R. Or do you think that's too restrictive?

Reply:
Unfortunately we would manually have manually pick the version of R and hope we are in sync with how "release" is defined in https://github.com/r-universe-org/actions/blob/v5/setup-r/action.yml. No binaries were downloaded when I tried "release" instead of "4.4" in https://staging.r-multiverse.org/apis. So I'm in favor of sticking with what the PR has now and then changing it later if we can. It will depend on our hosting options for the snapshot anyway. (p3m might be more accommodating than we expect.)

Reply:
To be clear, if we don't pick the right version of R that was considered "release" for Staging at the time, our snapshot may serve the wrong binaries. And it looks like we would have to parse https://github.com/r-universe-org/actions/blob/v5/setup-r/action.yml and make brittle assumptions to try to automatically detect what R version "release" actually is.

Reply:
Sorry I meant allow passing a specific version of R as the argument e.g. ‘4.4’. I’m in favour of something explicit like this so we don’t have to exactly follow R’s release schedule. To that extent perhaps we have a NULL default, so it only has effect if specified.

Reply:
Just added an `r_versions` argument for this.

Reply:
Thanks!